store/tikv/coprocessor.go

```
@@ -456,84 +474,156 @@ func (it *copIterator) Next() ([]byte, error) {
+	return nil, nil
 }
 
-// handleRegionErrorTask handles current task. It may be split into multiple tasks (in region split scenario).
-func (it *copIterator) handleRegionErrorTask(bo *Backoffer, task *copTask) []copResponse {
-	coprocessorCounter.WithLabelValues("rebuild_task").Inc()
-
-	newTasks, err := buildCopTasks(bo, it.store.regionCache, task.ranges, it.req.Desc)
-	if err != nil {
-		return []copResponse{{err: errors.Trace(err)}}
-	}
-	if newTasks == nil {
-		// TODO: check this, this should never happen.
-		return nil
+func buildCopTasksFromRemain(bo *Backoffer, cache *RegionCache, resp *coprocessor.Response, task *copTask, desc bool, streaming bool) ([]*copTask, error) {
+	remainedRanges := task.ranges
+	if streaming {
+		remainedRanges = calculateRemain(task.ranges, resp.Range, desc)
 	}
+	return buildCopTasks(bo, cache, remainedRanges, desc, streaming)
+}
 
-	ret := make([]copResponse, 0, len(newTasks))
-	for _, t := range newTasks {
-		resp := it.handleTask(bo, t)
-		ret = append(ret, resp...)
+func calculateRemain(ranges *copRanges, split *coprocessor.KeyRange, desc bool) *copRanges {
+	len := ranges.len()
+	if desc {
+		n := sort.Search(len, func(i int) bool {
+			r := ranges.at(i)
+			return bytes.Compare(split.End, r.StartKey) <= 0
+		})
+		if n == 0 {
+			log.Fatal("for any valid tikv response, n > 0 should always hold!")
+		}
+		r := ranges.at(n - 1)
+		ret := ranges.slice(0, n-1)
+		ret.last = &kv.KeyRange{r.StartKey, r.EndKey}
+		// [r.StartKey -- split.EndKey -- r.EndKey) => [r.StartKey -- split.EndKey)
+		if bytes.Compare(split.End, r.EndKey) < 0 {
+			ret.last.EndKey = split.End
+		}
+		return ret
 	}
-	return ret
+
+	key := split.Start
+	// Binary search for the range r that [r, r+1) is not consumed.
+	n := sort.Search(len, func(i int) bool {
+		r := ranges.at(i)
+		return bytes.Compare(r.EndKey, key) > 0
+	})
+	if n >= len {
+		log.Fatal("for any valid tikv response, n < len should always hold!")
```

Do not `log.Fatal`
just let it panic if n >= len.